File: cpp/visitor/visitor.cpp
#include <iostream>
#include <memory>

class Element;

class Visitor {
public:
    void process_element(Element &element) {
        std::cout << "Processing element!" << std::endl;
    }
};

class Element {
public:
    void accept_visitor(Visitor &visitor) {
        visitor.process_element(*this);
    }
};

int main() {
    Visitor visitor;
    Element element;
    element.accept_visitor(visitor);
    return 0;
}

File: python/visitor/visitor.py
class Visitor:
    def process_elemen(self, element: "Element"):
        print("Processing element!")


class Element:
    def accept_visitor(self, visitor: Visitor):
        visitor.process_elemen(self)


def main():
    element = Element()
    visitor = Visitor()
    element.accept_visitor(visitor)


if __name__ == '__main__':
    main()
